rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -61,6 +60,7 @@ object Reduce {
       private val urnMap: Map[String, Par]
   )(
       implicit
+      costAccountingAlg: CostAccounting[M],
```

Comment:
Just my 2 cents: this was NOT DONE this way on purpose. Interpreter should be decoupled from the `CostAccounting` logic. It should be layered on top of it using `ChargingReducer`.

Reply:
I figured, but the current implementation doesn't do that. It just adds a wrapper (ChargingReducer) around Reduce that delegates literally all of the calls to to Reduce, which handles the charging anyway. Reduce is parametric over M anyway, so it makes sense to pull state into the evaluation monad. That's far more parametric than an additional type-class.

Reply:
It wasn't possible to do it back then but it should be now. `ChargingReducer` would be delegating core logic to its normal counterpart but handle the cost logic by itself. This requires charging counterparts for Substitute and every function within Reducer though.

Reply:
This change is very close to code, easily-reversible, and it actually simplifies things. It might even not be problematic if/when we move cost accounting to a charging reducer. Anyway, yhx for the reminder about the intent behind it :)